test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,408 @@
+      "syscall",
+      "tpr_shadow",
+      "tsc",
+      "tsc_adjust",
+      "tsc_deadline_timer",
+      "tsc_known_freq",
+      "umip",
+      "vaes",
+      "vme",
+      "vmx",
+      "vnmi",
+      "vpclmulqdq",
+      "vpid",
+      "waitpkg",
+      "x2apic",
+      "xgetbv1",
+      "xsave",
+      "xsavec",
+      "xsaveopt",
+      "xsaves",
+      "xtopology"
+    ],
+    "model": "12th Gen Intel(R) Core(TM) i9-12900H"
+  },
+  "firmware": {
+    "bios": {
+      "date": "04/01/2014",
+      "vendor": "SeaBIOS",
+      "version": "1.15.0-1"
+    }
+  },
+  "hostname": "localhost.localdomain",
+  "nics": [
+    {
+      "ip": "192.168.222.122",
+      "mac": "00:60:2f:31:81:01",
+      "model": "0x1af4 0x0001",
+      "name": "enp1s0",
+      "pxe": true
+    },
+    {
+      "ip": "fe80::570a:edf2:a3a7:4eb8%enp1s0",
+      "mac": "00:60:2f:31:81:01",
+      "model": "0x1af4 0x0001",
+      "name": "enp1s0",
+      "pxe": true
+    }
+  ],
+  "ramMebibytes": 4096,
+  "storage": [
+    {
+      "name": "/dev/disk/by-path/pci-0000:04:00.0",
+      "rotational": true,
+      "sizeBytes": 21474836480,
+      "type": "HDD",
+      "vendor": "0x1af4"
+    }
+  ],
+  "systemVendor": {
+    "manufacturer": "QEMU",
+    "productName": "Standard PC (Q35 + ICH9, 2009)"
+  }
+}
+`
+
+var _ = Describe("BMO Upgrade", func() {
+	var (
+		specName               = "upgrade"
+		namespace              *corev1.Namespace
+		bmoNamespace           string
+		bmoDeployName          string
+		bmoReleaseBranch       string
+		bmcUser                string
+		bmcPassword            string
+		bmcAddress             string
+		bootMacAddress         string
+		bmoKustomization       string
+		upgradeClusterProvider bootstrap.ClusterProvider
+		upgradeClusterProxy    framework.ClusterProxy
+		bmh                    metal3api.BareMetalHost
+	)
+	BeforeEach(func() {
+		bmcUser = e2eConfig.GetVariable("BMC_USER")
+		bmcPassword = e2eConfig.GetVariable("BMC_PASSWORD")
+		bmcAddress = e2eConfig.GetVariable("BMC_ADDRESS")
+		bootMacAddress = e2eConfig.GetVariable("BOOT_MAC_ADDRESS")
+		bmoNamespace = "baremetal-operator-system"
+		bmoDeployName = "baremetal-operator-controller-manager"
+		bmoReleaseBranch = e2eConfig.GetVariable("BMO_RELEASE_BRANCH")
+		bmoUpgradeFromImageName := fmt.Sprintf("quay.io/metal3-io/baremetal-operator:%s", bmoReleaseBranch)
+		bmoKustomization = e2eConfig.GetVariable("BMO_KUSTOMIZATION")
+		bmoUpgradeFromKustomization := e2eConfig.GetVariable("BMO_UPGRADE_FROM_KUSTOMIZATION")
+
+		bmoUpgradeFromImage := clusterctl.ContainerImage{
+			Name:         bmoUpgradeFromImageName,
+			LoadBehavior: clusterctl.MustLoadImage,
+		}
+
+		if useExistingCluster {
+			k8sUpgradeContextName := e2eConfig.GetVariable("KUBECONFIG_UPGRADE_CONTEXT_NAME")
+			By(fmt.Sprintf("Using kubeconfig context: %s for upgrade tests", k8sUpgradeContextName))
+			err := SwitchKubeconfigContext(kubeconfigPath, k8sUpgradeContextName)
+			Expect(err).NotTo(HaveOccurred())
```

Comment:
Please check here also if the variable is set or not, and don't change the context if the variable is not set.

Reply:
I was a bit scared that if this is not set, then the same cluster will be used for both upgrade test and the other tests, which will basically screw up the whole thing without any error returned. But well, perhaps that's on the user xD